Add rollover log driver, and --log-driver-opts flag

wlan0 opened the pull request:
Closes #8911

This adds log rollover capability to docker. I have added a new driver called `rollover` for this.

There are two options that can be specified along with this option. In order to specify these options, I added a new flag called `--log-driver-opts`. That would allow one to specify options like this

`docker run --logging-driver rollover --log-driver-opts max_size=1k --log-driver-opts file_count=10`

cpuguy83 commented on daemon/logger/rollover/rollover.go:
```diff
@@ -0,0 +1,138 @@
+	return err
+}
+
+func writeLog(l *RolloverLogger) (int64, error) {
```
This probably needs to be protected by a mutex, no?

ibuildthecloud replied:
The impression I got was that calls to Log() are serialized.  @LK4D4 can you confirm that?

LK4D4 replied:
Uhm, serialized?

wlan0 replied:
I think he meant to ask if the calls to Log are being made in parallel. They are not, right
? My understanding was that each container has a copier that calls Log(). Therefore, we shouldn't worry about race conditions.

LK4D4 replied:
They are. You need to make logger threadsafe, because for now copier writes `stdout` and `stderr` concurrently.
Look #11472 :)

LK4D4 commented:
@wlan0 Gofmt check failed.
We need design review on new flag, but I sorta like it, not tried to use though.
ping @crosbymichael @tiborvass @jfrazelle

wlan0 commented:
@LK4D4 fixed the gofmt issue.

LK4D4 commented:
@wlan0 Haha, tests don't compile :)

wlan0 commented:
@LK4D4 Fixed it. There was a package that was unused, I removed it..

wlan0 commented:
@LK4D4 All tests pass now.. There were some unused packages earlier

tiborvass commented:
@wlan0 @LK4D4 I'd much rather have `--log-opt` like we have `--storage-opt`.

Also, I think we should not make it a new driver, but instead make the default `json-file` driver have options with this new flag. So rollover would be done like this: `docker run --log-opt max_file=10 --log-opt max_size=1k`.

Proposed defaults:
* max_size = 0 (no limit)
* max_file = 1

What do you guys think of this? If we can agree, we could move on to code review.

cpuguy83 commented:
+1 I like @tiborvass's suggestion.

LK4D4 commented:
@tiborvass I'm okay with moving that logic to default driver.

ibuildthecloud commented:
@tiborvass I also agree it should be in the default, we did it as a separate only because we didn't know if you guys would want us screwing with the default logging driver.  For example, just keep the default stupid simple because you know it works.

tiborvass commented:
@ibuildthecloud except this PR would introduce log options, and that's a whole new world :)

ibuildthecloud commented:
Okay.  To sum up.

* Change `--log-driver-opts` to `--log-opts`
* Remove "rollover" driver and put logic in json-file
* max_size = 0 should mean unlimited
* Defaults should be max_size=0, max_file=1

Question:

* Should it be max-size or max_size, which is more consistent with existing Docker?
* `docker logs` should read from all historical logs or just the latest?

tiborvass commented:
@ibuildthecloud 

flag would be singular: `--log-opt`.

Let's make it `max-size` to make it consistent with `-restart on-failure[:max-retry]`. Also I generally prefer `-` rather than `_`.

If `docker logs` outputs only the latest file, the default value of `--tail` would need to change to `latest` or something like that. I'm not sure what's the best solution for that, I'm still thinking.

wlan0 commented:
@LK4D4 @tiborvass updated the code according to comments. 

- Now, the flag is `--log-opt` 
- the options use `-` instead of `_`, so `max-file` and `max-size` are the options
- the options have been added to the json-file log driver
- the options have default values `max-file=1` and `max-size=0` (infinite)
- updated docs

tiborvass commented:
@wlan0 in your example `--log-driver json-file` is not necessary since it's the default.

I agree with everything you pointed out. Don't forget to update the API as well.

What's left to decide is what to do with `docker logs`... :S @LK4D4 @jfrazelle @icecrime

pires commented:
Great!

mahnunchik commented:
:+1:

gtonic commented:
+1

wlan0 commented:
@tiborvass Once the max is reached, I go back to file1

ibuildthecloud commented:
@wlan0 What do the filename look like?  When a container is destroyed, do all the log files get deleted?

wlan0 commented:
The logs are always written to the default log file - /var/lib/docker/containers/$CONTAINER_ID-json.log

Then the logs are rotated amongst 
/var/lib/docker/containers/$CONTAINER_ID-json.log-1
/var/lib/docker/containers/$CONTAINER_ID-json.log-2
/var/lib/docker/containers/$CONTAINER_ID-json.log....
/var/lib/docker/containers/$CONTAINER_ID-json.log-$(n-1)

where n = the maximum number of log files

I start by writing to the default log file. Then once this file reaches (max_size/n), I remove the last log file (/var/lib/docker/containers/$CONTAINER_ID-json.log-$(n-1)) and switch the contents of file-$(n-2) to file-$(n-1), and file-$(n-3) to file-$(n-2) and so on...

i.e. The higher the number on the log file name, the older the log is.

When a container is destroyed, the log files are deleted as well

defunctzombie commented:
Any reason why the default is unlimited? Just thinking about situations where a user has not configured docker and their systems will fill up with logs.

cpuguy83 commented:
@defunctzombie because that's how it currently works.

wlan0 commented:
@tiborvass for the `docker logs` command, we could add a new flag called index (i)

The index flag represents the file number that you want to tail. `i` would be a valid option only for the rollover log driver. it would have a default value of 0, meaning you tail the latest logs.

this way, a user can tail whichever one of the log files they want to.

unclejack commented:
If any kind of driver specific options are to be added for logging, these need to be specified in a dynamic way without adding a ton of flags which are driver dependent.

This would be OK:
```
docker logs --driver-options "option1=value1,option2=value2" container-id
```

NOT OK:
```
docker logs --some-driver-option 6 --some-driver-another-option foo container-id
```

tiborvass commented:
@unclejack we already have `--storage-opt`. So the format is `--log-opt k1=v1 --log-opt k2=v2`

tiborvass commented:
@wlan0 like I said earlier, there would not be a `rollover` driver, it would be part of the default json-file driver.

Everyone, are we all okay naming the different files `*.log then *.log-1, then *.log-2` ? Shouldn't we do `*.log then *.log.1 then *.log.2` instead?

@wlan0 if you want to add a new `--index` flag to `docker logs`, it has to have a meaning for ALL drivers, and I'm afraid it does not. So we need a better solution. Basically each driver would get a request saying "gimme the last `--tail` logs". By default `--tail=all`. Should `all` be interpreted differently by different drivers? I feel like there's something wrong with `docker logs`. Maybe `docker logs` only makes sense for with the default driver :S Ping @LK4D4 @icecrime @crosbymichael

LK4D4 commented:
I think that developing `docker logs` for each driver is PITA and doesn't worth it.
But this driver will be default, so for now we can end with tailing last file and document it.

tiborvass commented:
@LK4D4 then how would you tail ALL files ?

LK4D4 commented:
@tiborvass I would just document that you can't tail ALL files now. I'm not sure that there is person in world who want so big tails(I'm sure they exist, but they want it only to make me sad).

tiborvass commented:
@LK4D4 how about having `--tail last` by default, and still keep `--tail all` ?

LK4D4 commented:
@tiborvass Do we have `--tail=all` already implemented to keep it?

tiborvass commented:
@LK4D4 currently `--tail=all` is the default behavior :)

LK4D4 commented:
@tiborvass But currently there is no rollover :)

wlan0 commented:
@tiborvass -- I like this ` *.log, *.log.1 *.log.2 ` better too. I'll update the code.

tiborvass commented:
@LK4D4 yes but if `--log-opt max-size=1k --log-opt max-file=3` is set, then `--tail=all` should still mean all logs. So I would suggest `--tail=latest` be the default and return only logs from the latest file.

Also, `docker logs` should probably error out if the driver is something else than json-file.

ibuildthecloud commented:
I don't think you should rollover at max-size/n.  If I set max-file 5 and max-size as 50mb, in my brain it makes more sense that that would create 5 50mb files (=250mb).  That seems more in line with other logging systems.

For logs I think we should do one of the following:

1. Don't change the `logs` UI at all.  `docker logs --tail=all` means that we would pipe the response of `log-n`, `log-(n-1)`, ..., `log`.
2. If option 1 is difficult, we punt and just document that `docker logs` only works against the latest log.

tiborvass commented:
@ibuildthecloud I agree with you for max-file and max-size.

I think it makes sense that by default docker logs returns only the latest log file. But I don't want that to be upon `--tail=all` coz that would be misleading, hence have `--tail=all` return all logs, and make `--tail=latest` default that would only return the latest log file.

On second thought, each logging driver could implement however they want both `--tail=all` and `--tail=latest`, on a best effort basis.

wlan0 commented:
@tiborvass @ibuildthecloud 

I've accommodate the changes suggested. To sum up the current code behavior,

- Now, the flag is `--log-opt`
- the options use `-` instead of `_`, so `max-file` and `max-size` are the options
- the options have been added to the `json-file` log driver
- the options have default values `max-file=1` and `max-size=0` (infinite)
- updated docs
- we do `*.log then *.log.1 then *.log.2` instead of `*.log then *.log-1 then *.log-2` 
- we rollover at `max-size` for each file, instead of `max-size/n`


`docker logs`

- `docker logs --tail all` prints all the logs (from all files) in order in which they were written,
- `docker logs --tail latest` print the current log file only
- `--follow` option has the same behavior as before
- `default tail option is 'latest'`
- `docker logs --tail n`,  where n is a number tails the last n lines of the current log file only

tiborvass commented:
@wlan0 wow you rock! Just keep in mind this is still in design review and there can be back-and-forths :)

Small nit: I don't see why `docker logs --tail n` should tail the last n lines of the current log file only. It's fine to optimize for that's probably the most frequent usecase though. We could check the number of lines TailFile returned, and if it's less than what was requested, then move on to the next file.

@LK4D4 @ibuildthecloud @jfrazelle do you guys agree on `--tail=latest` vs `--tail=all` and make `--tail=latest` default?

ibuildthecloud commented:
@tiborvass I agree, with all and latest, and that latest should be the default.

wlan0 commented:
@tiborvass Fixed the code to return last n lines combining multiple files, if latest file doesn't have enough lines

wlan0 commented:
ping @tiborvass

jfrazelle commented:
I like `--log-opt` thanks! and I think `--tail latest,all` is sane

tiborvass commented:
I'd love to have @LK4D4's opinion before moving to code review.

LK4D4 commented:
@tiborvass Ok for me

tiborvass commented on api/client/logs.go:
```diff
@@ -16,7 +16,7 @@ func (cli *DockerCli) CmdLogs(args ...string) error {
 		follow = cmd.Bool([]string{"f", "-follow"}, false, "Follow log output")
 		times  = cmd.Bool([]string{"t", "-timestamps"}, false, "Show timestamps")
-		tail   = cmd.String([]string{"-tail"}, "all", "Number of lines to show from the end of the logs")
+		tail   = cmd.String([]string{"-tail"}, "latest", "Number of lines to show from the end of the logs")
```
add `latest` and `all` in the description. I suggest something like:
```
Last number of lines to show. Use `latest` (default) for latest file, or `all` for all logs.
```

@moxiegirl feel free to chime in

wlan0 replied:
@tiborvass Isn't there a limit (< 80) on the maximum length of command line flag descriptions?

tiborvass replied:
I'll leave it for docs people to decide once we move it to docs review

tiborvass replied:
ping @moxiegirl @SvenDowideit @thaJeztah how should we expose `latest` and `all` in the command line description ?

thaJeztah replied:
Sorry, missed this comment. Will have a look later and see if I can make a suggestion. The 80 chars limit can be a bit limiting; true

tiborvass commented on daemon/config.go:
```diff
@@ -83,7 +83,9 @@ func (config *Config) InstallFlags() {
-	flag.StringVar(&config.LogConfig.Type, []string{"-log-driver"}, "json-file", "Containers logging driver")
+	flag.StringVar(&config.LogConfig.Type, []string{"-log-driver"}, "json-file", "Logging driver")
+	config.LogConfig.Config = map[string]string{}
+	opts.MapVar(&config.LogConfig.Config, []string{"-log-opt"}, "Configure log driver")
```
to be more in sync with `--storage-opt`, I suggest `Set log driver options`.

ibuildthecloud commented:
@LK4D4 should we split this PR into two PRs?  That way we can get log-opts in first and and use it with the syslog driver.

LK4D4 commented:
@ibuildthecloud Yes, I'm okay with introducing `log-opts` for syslog first.

imton commented:
any idea when this will be merged? thx

tiborvass commented:
@wlan0 @ibuildthecloud is this thus dependent on #12422 ? If so, I would really appreciate if @wlan0 could close it and reopen it when it's ready again for review. Thanks!

wlan0 commented:
closing this, will reopen after #12422 gets merged

thaJeztah commented:
https://github.com/docker/docker/pull/12422 was merged. I'll re-open this PR so we don't forget

ping @wlan0 this probably needs a rebase now :)

wlan0 commented:
Thanks @thaJeztah! Will rebase now.

wlan0 commented:
@LK4D4 @tiborvass @thaJeztah rebased and ready.

thaJeztah commented:
@wlan0 awesome, thanks!

tiborvass commented on daemon/container.go:
```diff
@@ -950,10 +951,18 @@ func (container *Container) logPath(name string) (string, error) {
 	if err != nil {
 		return nil, err
 	}
+	if index != 0 {
```
`index > 0`, you never know :P

tiborvass commented on daemon/container.go:
```diff
@@ -1444,20 +1453,40 @@ func (container *Container) setupWorkingDirectory() error {
 
 func (container *Container) startLogging() error {
-	cfg := container.hostConfig.LogConfig
+	cfg := runconfig.LogConfig{}
```
```
cfg := runconfig.LogConfig{
    Type: container.hostConfig.LogConfig.Type,
    Config: container.mergeLogOpts(),
}
if cfg.Type == "" {
// ...
```

tiborvass commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -13,20 +16,50 @@ import (
+	if capacity != "" {
+		var multiplier int64
+		capString := capacity[:len(capacity)-1]
+		switch strings.ToLower(capacity[len(capacity)-1:]) {
```
Is there a reason why you cannot use `pkg/units` ?

wlan0 replied:
oh, didn't realize there's already code for that. I can use that.

wlan0 commented:
(repushed to resolve merge conflicts)

cpuguy83 commented:
Sadly this needs another rebase

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "rollover_log" git@github.com:wlan0/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

thaJeztah commented:
ping @LK4D4 @tiborvass @cpuguy83 PR was updated

wlan0 commented:
thanks @thaJeztah. Rebased and Repushed.

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -19,11 +22,12 @@ const (
+	f        *os.File   // store for closing
+	mu       sync.Mutex // protects buffer
+	capacity int64
+	n        int
```
What is `n` for?
I'm sure I can look and see, but would be helpful to know here in the struct definition.

wlan0 replied:
n stands for number of files.

tiborvass replied:
@wlan0 then would you mind adding a comment ?

wlan0 replied:
absolutely

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -38,36 +42,123 @@ func New(ctx logger.Context) (logger.Logger, error) {
 	}
+	var capval int64
+	capacity, ok := ctx.Config["max-size"]
+	if !ok {
```
Instead of doing the `if; else` can we just set `capval = -`; then do `if !ok { ..}`?

wlan0 replied:
you cant do that. I transform the data in the map to a different representation. The map holds string data like `1k (or) 1G`. I convert these into bytes.

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -38,36 +42,123 @@ func New(ctx logger.Context) (logger.Logger, error) {
+	var maxFiles int
+	maxFileString := ""
+	maxFileString, ok = ctx.Config["max-file"]
+	if !ok {
```
Same comment as `capval` above.

wlan0 replied:
same comment as above.

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -38,36 +42,123 @@ func New(ctx logger.Context) (logger.Logger, error) {
+		if err := rotate(name, l.n); err != nil {
+			return -1, err
+		}
+		os.Remove(name)
```
Do we need this since we are opening the file below with `os.O_TRUNC`?

cpuguy83 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -38,36 +42,123 @@ func New(ctx logger.Context) (logger.Logger, error) {
+			return err
+		}
+	}
+	if _, err := os.Stat(curr); os.IsNotExist(err) {
```
How come we're doing this at all?
Seems like `os.Rename` will handle this.

wlan0 replied:
os.Rename throws ENOENT if `oldpath` does not exist. That is why I have this.

cpuguy83 replied:
Oh you're right, I fat-fingered when I tested this

tiborvass commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -38,36 +42,123 @@ func New(ctx logger.Context) (logger.Logger, error) {
 	if err != nil {
 		return nil, err
 	}
+	var capval int64
```
```
var capval int64 = -1
//...
if ok {
  capval = ...
}
```

wlan0 commented:
@tiborvass @cpuguy83 Updated based on your comments.

cpuguy83 commented on runconfig/parse.go:
```diff
@@ -402,7 +402,14 @@ func parseLoggingOpts(loggingDriver string, loggingOpts []string) (map[string]st
 		return map[string]string{}, fmt.Errorf("Invalid logging opts for driver %s", loggingDriver)
 	}
-	//TODO - validation step
+	for key := range loggingOptsMap {
```
This should be handled by the logging driver and not here, imo.

wlan0 replied:
Thanks, I understand why. I'll change this.

cpuguy83 commented:
Functionality is generally working, however.
If I pass in, let's say... `--log-opt max-file=3` to the daemon, and then do `--log-opt max-size=10K` on `docker run`, the daemon `max-file` is not taken into account.

wlan0 commented:
@cpuguy83 The behavior you described is correct. It doesn't make sense to specify `max-file` without specifying `max-size`. With the current logic, it selects either the defaultlogConfig or runConfig.hostConfig.LogConfig, it doesn't merge them into one.

wlan0 commented:
ping @cpuguy83. The log-opts validations happens in the driver now.

cpuguy83 commented on opts/opts.go:
```diff
@@ -177,7 +177,7 @@ type ValidatorFctType func(val string) (string, error)
 
 func ValidateLogOpts(val string) (string, error) {
-	allowedKeys := map[string]string{}
+	allowedKeys := map[string]string{"max-file": "1", "max-size": "2"}
```
Looks like this got left here.

wlan0 replied:
This validates log-opts for the daemon. The log drivers are called only while creating a container. A validation step is needed at this point.

cpuguy83 replied:
hmm, I really don't like this.
Could we just instantiate a logger and call a ValidateOpts (after adding this to the interface) on it?
This is kind of a weird one since the values don't get used right away.

Anyone else want to weigh in?

LK4D4 replied:
Yes, I think it will be hard to track all options.

cpuguy83 replied:
ping @wlan0 Do you want to make this change.

wlan0 replied:
yes absolutely.

cpuguy83 commented:
@wlan0 Would you like someone to carry?

wlan0 committed:
add support for maximum log size, and max number of log files

Signed-off-by: wlan0 <sidharthamn@gmail.com>

wlan0 commented:
@cpuguy83 Updated the PR.

cpuguy83 commented:
LGTM

LK4D4 commented on daemon/logger/jsonfilelog/jsonfilelog.go:
```diff
@@ -59,15 +86,98 @@ func (l *JSONFileLogger) Log(msg *logger.Message) error {
+	return nil
+}
+
+func (l *JSONFileLogger) GetReaderByIndex(index int) (io.Reader, error) {
```
Omg, I don't like this so much. But I think we have no choice now. `GetReader` was superbad idea in the first place.
I don't know, maybe use type assertion instead of bringing this to interface?

cpuguy83 replied:
I didn't like it either...but couldn't think of anything better.
I don't like asserting to a specific type for this either.

Maybe we split the reader interface out of the main logger interface, since not everything will be implementing reader anyway.

Then in daemon/logs.go we can do type assertion to a reader, if that fails we return ErrNotSupported (or whatever the error is)

wlan0 replied:
I agree with cpuguy83. Doing a type assertion would just move the problem around. I don't like this either, but couldn't think of a better solution that doesn't involve extensive change.

LK4D4 replied:
`GetReader` is abomination which must die anyway. It's very sad, that now we have so much changes which rely on it. That's why I thought it was better to have type assertion, so no-one will try to adapt awful `json-file` logs retrieval.

wlan0 commented:
@cpuguy83 @LK4D4 I separated logger and reader interfaces.

jfrazelle commented:
oh man needs rebase then I promise we should be able to get this in, hopefully, no promises, well kinda... ;)

wlan0 commented:
not a problem :) Gimme till the end of the day.

wlan0 commented:
@jfrazelle the rebase was fortunately easier than last time. :) It's ready.

jfrazelle commented:
ping @LK4D4 for last review :)

calavera commented:
code LGMT. Moving to docs review.

calavera commented on docs/reference/logging/index.md:
```diff
@@ -17,7 +17,7 @@ container's logging driver. The following options are supported:
 | `none`      | Disables any logging for the container. `docker logs` won't be available with this driver.                                    |
 |-------------|-------------------------------------------------------------------------------------------------------------------------------|
-| `json-file` | Default logging driver for Docker. Writes JSON messages to file.  No logging options are supported for this driver.           |
+| `json-file` | Default logging driver for Docker. Writes JSON messages to file.  two logging options are supported for this driver.          |
```
`s/two/Two/`, I think there is also an extra white space before the beginning of the sentence that we can remove.

moxiegirl replied:
Default Docker logging driver. Writes JSON messages to a file. This driver supports two options.

wlan0 commented:
@calavera Fixed the doc!

moxiegirl commented on docs/reference/commandline/logs.md:
```diff
@@ -19,6 +19,7 @@ weight=1
       --since=""                Show logs since timestamp
       -t, --timestamps=false    Show timestamps
       --tail="all"              Number of lines to show from the end of the logs
+      --log-opt=""              Specify options to the logger, Currently supported only by json-file log driver
```
The entire command has a note that says it only works with the json-file log driver --- so you don't need to repeat that on the flag.

"Specify logger options. "  is tighter

moxiegirl commented on docs/reference/commandline/logs.md:
```diff
@@ -29,7 +30,7 @@ The `docker logs --follow` command will continue streaming the new output from
 
 Passing a negative number or a non-integer to `--tail` is invalid and the
-value is set to `all` in that case. This behavior may change in the future.
+value is set to `latest` in that case. This behavior may change in the future.
```
Omit "This behavior may change in the future" as the facts of that statement are inherent.

moxiegirl commented on docs/reference/logging/index.md:
```diff
@@ -25,6 +25,16 @@ container's logging driver. The following options are supported:
+   --log-opt max-file=[0-9+]
+
+`max-size` sets the size of each of the log files that are rolled over
+`max-file` sets the maximum number of log files that are rolled over amongst
```
missing something after "amongst" --- also --- what is the point of the max size?  and 9+ does that mean unlimited?  Why aren't we use date or something --- roll over files that are aged past 9 days and x size...

---> Something like

### The json-file options

The `json-file` logging driver supports the following options:

```
--log-opt max-size=[0-9+][k/m/g/K/M/G]
--log-opt max-file=[0-9+]
```

`max-size` Logs that reach this size are rolled over.  You can set the size in kilobytes (k), megbytes (m), or gigbytes (g).

wlan0 replied:
@moxiegirl I am trying to denote that one or more digits should be present before the unit field `[k | m | g]`. 

The default value for both these options are empty string. How about something like this?

### The json-file options

The following logging options are supported for the `json-file` logging driver:
```
   --log-opt max-size=""
   --log-opt max-file=""
```
`max-size` Logs that reach this size are rolled over. You can set the size in kilobytes(k), megabytes(m), or gigabytes(g). eg `--log-opt max-size=50m`
`max-file` Maximum number of files that a log is rolled over before being discarded. eg `--log-opt max-file=100`

moxiegirl commented:
@wlan0 Thanks for the contribution. Please check your docs output with a `make docs` from the `docs` dir.

thaJeztah commented on docs/reference/logging/index.md:
```diff
@@ -17,7 +17,7 @@ container's logging driver. The following options are supported:
 | `none`      | Disables any logging for the container. `docker logs` won't be available with this driver.                                    |
 |-------------|-------------------------------------------------------------------------------------------------------------------------------|
-| `json-file` | Default logging driver for Docker. Writes JSON messages to file.  No logging options are supported for this driver.           |
+| `json-file` | Default logging driver for Docker. Writes JSON messages to file. Two logging options are supported for this driver.           |
```
I think you can remove "Two logging options are supported for this driver.". We don't mention this for the other drivers and it is explained below, so doesn't add much.

wlan0 commented:
@moxiegirl @thaJeztah I've updated the docs based on your comments.

thaJeztah commented on docs/reference/logging/index.md:
```diff
@@ -25,6 +25,16 @@ container's logging driver. The following options are supported:
+The following logging options are supported for the `json-file` logging driver:
+
+   --log-opt max-size=(unlimited)
+   --log-opt max-file=1
```
This isn't rendered as code-block. Perhaps the indentation is 3 in stead of 4 spaces? You can test the docs by running `make docs`, which will build the and serve the docs in a container so that you can preview them in your webbrowser.

thaJeztah replied:
Thinking of the `max-size` option; if the default is "unlimited", then specifying `max-files` **without** specifying `max-size` is a no-op? I.e., the max-size is never reached, so there will never be a roll-over?

wlan0 replied:
yes. That is correct. `max-file` is meaningless without max-size. It won't be honored even if you set it. If you want me to describe it differently, Im good with that.

thaJeztah commented on docs/reference/logging/index.md:
```diff
@@ -25,6 +25,16 @@ container's logging driver. The following options are supported:
 
+### The json-file options
+
+The following logging options are supported for the `json-file` logging driver:
```
Can you change this line to the example that @moxiegirl gave in https://github.com/docker/docker/pull/11485#discussion_r34721477 ?

wlan0 replied:
@thaJeztah the description written by @moxiegirl was my previous description. It seems to be a little ambiguous. I felt like this was clearer. What do you think?

wlan0 commented:
@thaJeztah @moxiegirl Updated.

draghuram commented:
Hi, 

Thanks for the PR as I have no doubt it adds a very useful feature to Docker. Here are some comments about the doc changes.

- `docs/reference/logging/index.md`

  - As @thaJeztah mentioned, I think it is worthwhile to clearly document that `max-file` is meaningless without `max-size`. At the same time, we should also mention the default value for `max-size` (which is `unlimited`). 

  - There is a missing space in "The `docker logs`command"  though it is not this PR's doing. 

- `docs/reference/commandline/logs.md`

  - It is not clear to me why `log-opt` option is added to `docker logs` command. What exactly can I pass with this option? I may be missing something here but the options `max-size` and `max-file` make sense only  for `docker run` command.  

  - If I run `docker logs` without any options, does it combine all the log files or just return the latest? I ask because the command seems to default to "latest" when an invalid value is passed to `--tail` option. Similarly, does `--since` option apply across all the log files? 

- `docs/reference/commandline/run.md`

  It would be beneficial if an example is provided in this file covering the new log options.

moxiegirl commented:
@draghuram makes very good points. @wlan0 can you clarify in the docs these points please? Thank you for being so cooperative on making changes. User would appreciate this extra clarification.

wlan0 commented:
@moxiegirl @draghuram Thanks! Absolutely not a problem. I would like to be as clear as possible while describing these options as well. 

Firstly, I've updated the docs keeping in mind all of your comments. Secondly, to answer your questions - 

1. `--log-opt` doesn't make sense in `logs` command. Good catch. Removed this from the `logs.md` file in the docs.
2.  If you run `docker logs` without options, it only provides the latest logs. Added this to the docs.

I've updated the docs according to the rest of your comments.

draghuram commented on docs/reference/logging/index.md:
```diff
@@ -25,6 +25,20 @@ container's logging driver. The following options are supported:
+
+`max-file` specifies the maximum number of files that a log is rolled over before being discarded. eg `--log-opt max-file=100`. If `max-size` is not set, then `max-file` is not honored.
+
+If `max-size` and `max-file` are set, `docker logs` only returns the log lines from the newest log file. 
```
I think this information really belongs in `logs.md`. In fact, I would like `logs.md` to provide some description of rollover feature and how that impacts the behavior of options. For example, `--tail`` description says that when an invalid value is passed, the value is set to `latest`. However, you can't tell what `latest` means just by looking at the man page.

draghuram replied:
Just to be clear, my comment above only applies to the line "If `max-size` and `max-file` are set, `docker logs` only returns the log lines from the newest log file.".

moxiegirl replied:
@draghuram I think you have a point about the logs.md file but it is beyond the scope of this PR. I created an issue here: https://github.com/docker/docker/issues/14711 and assigned it to myself.

draghuram commented:
@wlan0 Thanks for the changes. In addition to my comment above, I would really like to see an example in `run.md` but I am ok with the PR being merged without one.

moxiegirl commented:
LGTM thanks @wlan0

moxiegirl commented:
@thaJeztah for the merge.

thaJeztah commented:
LGTM! Thanks so much for being so patient and cooperative, @wlan0. Apologies that it took so long :-)

wlan0 commented:
yaay finally!!! thanks all... :)

cdancy commented:
+1!!!

henfri commented:
Hello,

thanks for this feature!
Is it possible to use it with docker-compose?
How do I specify/call it in the compose.yml

Greetings,
Hendrik

thaJeztah commented:
@henfri please don't comment on closed/merged PR's for support questions; the issue tracker is meant for tracking bugs, and reviewing pull requests; your question is better asked;

- in the #docker or #docker-compose) IRC channel on freenode
- the docker-user Google Group (https://groups.google.com/forum/m/#!forum/docker-user
- https://forums.docker.com
